File: e2e/java/src/test/java/dev/kreuzberg/e2e/SitemapTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/** E2e tests for category: sitemap. */
class SitemapTest {
    @Test
    void testSitemapBasic() {
        // Parses a standard urlset sitemap
        var result = Kreuzcrawl.scrape();
        assertEquals(4, result.urls().size());
        assertEquals(true, result.has_lastmod());
    }

    @Test
    void testSitemapCompressedGzip() {
        // Parses a gzip-compressed sitemap file
        var result = Kreuzcrawl.scrape();
        assertEquals(3, result.urls().size());
    }

    @Test
    void testSitemapEmpty() {
        // Handles empty sitemap gracefully
        var result = Kreuzcrawl.scrape();
        assertEquals(0, result.urls().size());
    }

    @Test
    void testSitemapFromRobotsTxt() {
        // Discovers sitemap via robots.txt Sitemap directive
        var result = Kreuzcrawl.scrape();
        assertEquals(4, result.urls().size());
    }

    @Test
    void testSitemapIndex() {
        // Follows sitemap index to discover child sitemaps
        var result = Kreuzcrawl.scrape();
        assertEquals(3, result.urls().size());
    }

    @Test
    void testSitemapLastmodFilter() {
        // Filters sitemap URLs by lastmod date
        var result = Kreuzcrawl.scrape();
        assertEquals(4, result.urls().size());
        assertEquals(true, result.has_lastmod());
    }

    @Test
    void testSitemapOnlyMode() {
        // Uses sitemap URLs exclusively without following page links
        var result = Kreuzcrawl.scrape();
        assertEquals(4, result.urls().size());
    }

    @Test
    void testSitemapXhtmlLinks() {
        // Parses sitemap with XHTML namespace alternate links
        var result = Kreuzcrawl.scrape();
        assertEquals(2, result.urls().size());
        assertEquals(false, result.has_lastmod());
    }

}

File: e2e/go/sitemap_test.go
// E2e tests for category: sitemap
package e2e_test

import (
	"strings"
	"testing"

	pkg "github.com/kreuzberg-dev/kreuzcrawl"
)

func Test_SitemapBasic(t *testing.T) {
	// Parses a standard urlset sitemap
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Urls) != 4 {
		t.Errorf("equals mismatch: got %q", len(result.Urls))
	}
	if result.HasLastmod != true {
		t.Errorf("equals mismatch: got %q", result.HasLastmod)
	}
}

func Test_SitemapCompressedGzip(t *testing.T) {
	// Parses a gzip-compressed sitemap file
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Urls) != 3 {
		t.Errorf("equals mismatch: got %q", len(result.Urls))
	}
}

func Test_SitemapEmpty(t *testing.T) {
	// Handles empty sitemap gracefully
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Urls) != 0 {
		t.Errorf("equals mismatch: got %q", len(result.Urls))
	}
}

func Test_SitemapFromRobotsTxt(t *testing.T) {
	// Discovers sitemap via robots.txt Sitemap directive
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Urls) != 4 {
		t.Errorf("equals mismatch: got %q", len(result.Urls))
	}
}

func Test_SitemapIndex(t *testing.T) {
	// Follows sitemap index to discover child sitemaps
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Urls) != 3 {
		t.Errorf("equals mismatch: got %q", len(result.Urls))
	}
}

func Test_SitemapLastmodFilter(t *testing.T) {
	// Filters sitemap URLs by lastmod date
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Urls) != 4 {
		t.Errorf("equals mismatch: got %q", len(result.Urls))
	}
	if result.HasLastmod != true {
		t.Errorf("equals mismatch: got %q", result.HasLastmod)
	}
}

func Test_SitemapOnlyMode(t *testing.T) {
	// Uses sitemap URLs exclusively without following page links
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Urls) != 4 {
		t.Errorf("equals mismatch: got %q", len(result.Urls))
	}
}

func Test_SitemapXhtmlLinks(t *testing.T) {
	// Parses sitemap with XHTML namespace alternate links
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Urls) != 2 {
		t.Errorf("equals mismatch: got %q", len(result.Urls))
	}
	if result.HasLastmod != false {
		t.Errorf("equals mismatch: got %q", result.HasLastmod)
	}
}
